pandas/core/arrays/string_.py

```diff
@@ -79,37 +84,114 @@ class StringDtype(ExtensionDtype):
     StringDtype
     """
 
-    name = "string"
-
     #: StringDtype.na_value uses pandas.NA
     na_value = libmissing.NA
+    _metadata = ("storage",)
+
+    def __init__(self, storage=None):
+        if storage is None:
+            storage = get_option("mode.string_storage")
+        if storage not in {"python", "pyarrow"}:
+            raise ValueError(
+                f"Storage must be 'python' or 'pyarrow'. Got {storage} instead."
+            )
+        self.storage = storage
+
+    @property
+    def name(self):
+        return f"StringDtype[{self.storage}]"
 
     @property
     def type(self) -> Type[str]:
         return str
 
     @classmethod
-    def construct_array_type(cls) -> Type[StringArray]:
+    def construct_from_string(cls, string):
+        """
+        Construct a StringDtype from a string.
+
+        Parameters
+        ----------
+        string : str
+            The type of the name. The storage type will be taking from `string`.
+            Valid options and their storage types are
+
+            ========================== ==============
+            string                     result storage
+            ========================== ==============
+            ``'string'``               global default
+            ``'string[python]'``       python
+            ``'StringDtype[python]'``  python
+            ``'string[pyarrow]'``      pyarrow
+            ``'StringDtype[pyarrow]'`` pyarrow
+            ========================== =============
+
+        Returns
+        -------
+        StringDtype
+
+        Raise
+        -----
+        TypeError
+            If the string is not a valid option.
+
+        """
+        if not isinstance(string, str):
+            raise TypeError(
+                f"'construct_from_string' expects a string, got {type(string)}"
+            )
+        if string == "string":
+            # TODO: use global default
+            return cls()
+        elif string in {"string[python]", "StringDtype[python]"}:
+            return cls(storage="python")
+        elif string in {"string[pyarrow]", "StringDtype[pyarrow]"}:
+            return cls(storage="pyarrow")
+        else:
+            raise TypeError(f"Cannot construct a '{cls.__name__}' from '{string}'")
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, str) and other == "string":
+            return True
+        return super().__eq__(other)
```

Comment:
What is actually the behaviour here? Are "string[python]" and "string[arrow]" seen as not equal, I suppose?

Reply:
That's correct. Not sure what the behavior should be if we delay looking up the storage. that could complicate things so will leave for a follow-on after better understanding of what we want, and when the lookup would be final.

```
>>> with pd.option_context("string_storage", "python"):
...     print(StringDtype() == "string")
... 
True
>>> with pd.option_context("string_storage", "pyarrow"):
...     print(StringDtype() == "string")
... 
True
>>> with pd.option_context("string_storage", "python"):
...     print(StringDtype() == "string[python]")
... 
True
>>> with pd.option_context("string_storage", "python"):
...     print(StringDtype() == "string[pyarrow]")
... 
False
>>> with pd.option_context("string_storage", "pyarrow"):
...     print(StringDtype() == "string[python]")
... 
False
>>> with pd.option_context("string_storage", "pyarrow"):
...     print(StringDtype() == "string[pyarrow]")
... 
True
>>> StringDtype(storage="python") == "string"
True
>>> 
>>> StringDtype(storage="pyarrow") == "string"
True
>>> 
>>> StringDtype(storage="python") == StringDtype(storage="python")
True
>>> 
>>> StringDtype(storage="pyarrow") == StringDtype(storage="pyarrow")
True
>>> 
>>> StringDtype(storage="python") == StringDtype(storage="pyarrow")
False
>>> 
>>> StringDtype(storage="pyarrow") == "string[pyarrow]"
True
>>> 
>>> StringDtype(storage="pyarrow") == "string[python]"
False
```

Reply:
i think this is ok, as we consider dtypes to be exactly equal. later on maybe we can allow equivalent tests (possibly open an issue about this)